gateway-ha/src/main/java/io/trino/gateway/proxyserver/ProxyRequestHandler.java

```diff
@@ -169,13 +172,19 @@ private void performRequest(
 
         FluentFuture<ProxyResponse> future = executeHttp(request);
 
+        List<NewCookie> sessionCookies = new ArrayList<>();
         if (statementPaths.stream().anyMatch(request.getUri().getPath()::startsWith) && request.getMethod().equals(HttpMethod.POST)) {
             future = future.transform(response -> recordBackendForQueryId(request, response), executor);
+            if (includeClusterInfoInResponse) {
+                // add trinoClusterHost to sessionCookies
+                sessionCookies.add(new NewCookie.Builder("trinoClusterHost").value(remoteUri.getHost()).build());
+            }
         }
 
+        ImmutableList<NewCookie> finalSessionCookie = ImmutableList.copyOf(sessionCookies);
         setupAsyncResponse(
                 asyncResponse,
-                future.transform(response -> buildResponse(response, oauth2GatewayCookie), executor)
+                future.transform(response -> buildResponse(response, oauth2GatewayCookie, finalSessionCookie), executor)
```

Comment:
This is a bit more awkward than it needs to be because `getOAuth2GatewayCookie` returns an `ImmutableList`. Can you change it's return type to `List` and update the method to use some mutable implementation like `ArrayList`, then add your `trinoClusterHost` cookie to that list? Then you can revert the changes to `buildResponse`